kong/runloop/wasm.lua

```diff
@@ -1,141 +1,537 @@
+
+  elseif typ == TYPE_COMBINED and service_id and route_id then
+    local routes = state.combined[service_id]
+
+    if not routes then
+      routes = {}
+      state.combined[service_id] = routes
+    end
 
-  for chain, err in kong.db.wasm_filter_chains:each() do
+    routes[route_id] = ref
+  end
+end
+
+
+---
+-- Build a combined filter list from 1-2 filter chain entities.
+--
+-- Disabled filter chains are skipped, and disabled filters are
+-- skipped.
+--
+-- Returns `nil` if no enabled filters are found.
+--
+---@param service_chain? kong.db.schema.entities.wasm_filter_chain
+---@param route_chain?   kong.db.schema.entities.wasm_filter_chain
+---
+---@return kong.db.schema.entities.wasm_filter[]?
+local function build_filter_list(service_chain, route_chain)
+  ---@type kong.db.schema.entities.wasm_filter[]|nil
+  local combined
+  local n = 0
+
+  if service_chain and service_chain.enabled then
+    for _, filter in ipairs(service_chain.filters) do
+      if filter.enabled then
+        n = n + 1
+        combined = combined or {}
+        combined[n] = filter
+      end
+    end
+  end
+
+  if route_chain and route_chain.enabled then
+    for _, filter in ipairs(route_chain.filters) do
+      if filter.enabled then
+        n = n + 1
+        combined = combined or {}
+        combined[n] = filter
+      end
+    end
+  end
+
+  return combined
+end
+
+
+---
+-- Unconditionally rebuild and return a new wasm state table from the db.
+--
+---@param  db                       table # kong.db
+---@param  version                  any
+---@param  old_state                kong.runloop.wasm.state
+---@return kong.runloop.wasm.state? new_state
+---@return string?                  err
+local function rebuild_state(db, version, old_state)
+  local route_chains = {}
+  local service_chains_by_id = {}
+
+  ---@type kong.runloop.wasm.state
+  local state = {
+    by_service = {},
+    by_route = {},
+    combined = {},
+    version = version,
+  }
+
+  local all_chains = {}
+  local page_size = db.wasm_filter_chains.max_page_size
+
+  for chain, err in db.wasm_filter_chains:each(page_size) do
     if err then
-      error(err)
+      return nil, "failed iterating filter chains: " .. tostring(err)
     end
 
-    local service_id = chain.service and chain.service.id
-    local route_id = chain.route and chain.route.id
+    if chain.enabled then
+      local route_id = chain.route and chain.route.id
+      local service_id = chain.service and chain.service.id
 
-    -- global filter chains first
-    if not service_id and not route_id then
-      extend(filters, chain.filters)
+      local chain_type = service_id and TYPE_SERVICE or TYPE_ROUTE
 
-    -- service filter chains
-    elseif service_id and not route_id then
-      if service_id == service.id then
-        extend(filters, chain.filters)
+      insert(all_chains, {
+        type       = chain_type,
+        service    = (chain_type == TYPE_SERVICE and chain) or nil,
```

Comment:
```suggestion
        service_chain    = (chain_type == TYPE_SERVICE and chain) or nil,
```
or `service_level_chain`, which would have made the code easier to read for me. Not a blocker.